data/transactions/logic/assembler.go

```diff
@@ -1182,13 +1223,359 @@ const AssemblerDefaultVersion = 1
+// preexisting intcblocks in the code.
+func (ops *OpStream) optimizeIntcBlock() error {
+	type intFrequency struct {
+		value uint64
+		freq  int
+	}
+
+	if ops.hasIntcBlock {
+		// don't optimize an existing intcblock, only int pseudo-ops
+		return nil
+	}
+
+	freqs := make([]intFrequency, len(ops.intc))
+
+	for i, value := range ops.intc {
+		freqs[i].value = value
+	}
+
+	for _, ref := range ops.intcRefs {
+		found := false
+		for i := range freqs {
+			if ref.value == freqs[i].value {
+				freqs[i].freq++
+				found = true
+				break
+			}
+		}
+		if !found {
+			return ops.errorf("Value not found in intc block: %d", ref.value)
+		}
+	}
+
+	for _, f := range freqs {
+		if f.freq == 0 {
+			return ops.errorf("Member of intc block is not used: %d", f.value)
+		}
+	}
+
+	// sort values by greatest to smallest frequency
+	// since we're using a stable sort, constants with the same frequency
+	// will retain their current ordering (i.e. first referenced, first in int block)
+	sort.SliceStable(freqs, func(i, j int) bool {
+		return freqs[i].freq > freqs[j].freq
+	})
+
+	// sort refs from last to first in raw
+	// this way when we iterate through them and potentially change the size of raw,
+	// the later positions will not affect the indexes of the earlier positions
+	sort.Slice(ops.intcRefs, func(i, j int) bool {
+		return ops.intcRefs[i].position > ops.intcRefs[j].position
+	})
+
+	opIntc0 := OpsByName[ops.Version]["intc_0"].Opcode
+	opIntc1 := OpsByName[ops.Version]["intc_1"].Opcode
+	opIntc2 := OpsByName[ops.Version]["intc_2"].Opcode
+	opIntc3 := OpsByName[ops.Version]["intc_3"].Opcode
+	opIntc := OpsByName[ops.Version]["intc"].Opcode
+	opPushInt := OpsByName[ops.Version]["pushint"].Opcode
+
+	var scratch [binary.MaxVarintLen64]byte
+	raw := ops.pending.Bytes()
+	for _, ref := range ops.intcRefs {
+		singleton := false
+		newIndex := -1
+		for i, f := range freqs {
+			if ref.value == f.value {
+				singleton = f.freq == 1
+				newIndex = i
+				break
+			}
+		}
+		if newIndex == -1 {
+			return ops.errorf("Value not found in intc block: %d", ref.value)
+		}
+
+		var newBytes []byte
+		if singleton {
+			vlen := binary.PutUvarint(scratch[:], ref.value)
+			newBytes = make([]byte, 1+vlen)
+			newBytes[0] = opPushInt
+			copy(newBytes[1:], scratch[:vlen])
+		} else {
+			switch newIndex {
+			case 0:
+				newBytes = []byte{opIntc0}
+			case 1:
+				newBytes = []byte{opIntc1}
+			case 2:
+				newBytes = []byte{opIntc2}
+			case 3:
+				newBytes = []byte{opIntc3}
+			default:
+				newBytes = []byte{opIntc, uint8(newIndex)}
+			}
+		}
+
+		var currentBytesLen int
+		switch raw[ref.position] {
+		case opIntc0, opIntc1, opIntc2, opIntc3:
+			currentBytesLen = 1
+		case opIntc:
+			currentBytesLen = 2
+		default:
+			return ops.errorf("Unexpected op at intReference: %d, from source line %d", raw[ref.position], ops.OffsetToLine[ref.position])
+		}
+
+		raw = replaceBytes(raw, ref.position, currentBytesLen, newBytes)
```

Comment:
Since currentBytesLen and newBytes will so often be the same length, I wonder if your replaceBytes should relax its rule that it always produces a new slice, and instead return the existing slice with the bytes replaced?

In fact, when len(newBytes) < currentBytesLen, it can also return in place. It feels in keeping with "append" which has a similar contract.

Just bike shedding.  No need to do this.

Reply:
Great suggestion, I'm happy to make this improvement in a follow-up PR

Reply:
I managed to implement this in the most recent commit